File: debut-with-a-catamorphism/MyOption.scala

// An excercise from Tony Morris.
// See http://blog.tmorris.net/debut-with-a-catamorphism/

trait MyOption[+A] {
  // single abstract method
  def cata[X](some: A => X, none: => X): X

  private def none[A] = new MyOption[A] {
    def cata[X](s: A => X, n: => X) = n
  }

  private def some[A](a: A) = new MyOption[A] {
    def cata[X](s: A => X, n: => X) = s(a)
  }

  private def identity[A](a: A) = a

  def map[B](f: A => B): MyOption[B] = cata(f andThen some, none)

  def flatMap[B](f: A => MyOption[B]): MyOption[B] = cata(f, none)

  def getOrElse[AA >: A](e: => AA): AA = cata(identity[A], e)

  def filter(p: A => Boolean): MyOption[A] = cata(a => if (p(a)) some(a) else none, none)

  def foreach(f: A => Unit): Unit = cata(f, ())

  def isDefined: Boolean = cata(_ => true, false)

  def isEmpty: Boolean = !isDefined

  // WARNING: not defined for None
  def get: A = cata(identity[A], error("None.get"))

  def orElse[AA >: A](o: MyOption[AA]): MyOption[AA] = cata(_ => this, o)

  def toLeft[X](right: => X): Either[A, X] = cata(Left(_), Right(right))

  def toRight[X](left: => X): Either[X, A] = cata(Right(_), Left(left))

  def toList: List[A] = cata(List(_), Nil)

  def iterator: Iterator[A] = cata(Iterator(_), Iterator())

  override def toString: String = cata("Some(" + _ + ")", "None")
}

object MyOption {
  def none[A] = new MyOption[A] {
    def cata[X](s: A => X, n: => X) = n
  }

  def some[A](a: A) = new MyOption[A] {
    def cata[X](s: A => X, n: => X) = s(a)
  }
}

File: debut-with-a-catamorphism/MyOption.hs
{-# LANGUAGE RankNTypes #-}
 
-- Data.Maybe
newtype Option a = Option { cata :: forall x. (a -> x) -> x -> x }
 
-- Just
some :: a -> Option a
some a = Option (\s _ -> s a)
 
-- Nothing
none :: Option a
none = Option (const id)
 
-- fmap
map' :: (a -> b) -> Option a -> Option b
map' f m = cata m (some . f) none
 
-- (>>=)
flatMap :: (a -> Option b) -> Option a -> Option b
flatMap f m = cata m f none
 
-- fromMaybe
getOrElse :: Option a -> a -> a
getOrElse m a = cata m id a
 
filter :: Option a -> (a -> Bool) -> Option a
filter m p = cata m (\a -> if p(a) then m else none) none
 
-- mapM_
foreach :: Option a -> (a -> IO ()) -> IO ()
foreach m f = cata m f (return ())
 
-- isJust
isDefined :: Option a -> Bool
isDefined m = cata m (const True) False
 
-- isNothing
isEmpty :: Option a -> Bool
isEmpty m = not $ isDefined m
 
-- WARNING: not defined for None
-- fromJust
get :: Option a -> a
get m = cata m id (error "None.get")

-- mplus
orElse :: Option a -> Option a -> Option a
orElse m n = cata m (const m) n
 
toLeft :: Option a -> x -> Either a x
toLeft m x = cata m Left (Right x)
 
toRight :: Option a -> x -> Either x a
toRight m x = cata m Right (Left x)
 
-- maybeToList
toList :: Option a -> [a]
toList m = cata m (\a -> [a]) []
 
iterator = error "bzzt. This is Haskell silly."

toMaybe m = cata m (\a -> Just a) Nothing
